docs/Testing.fsx

```diff
@@ -947,4 +956,174 @@ qHisto
 
 (***hide***)
 qHisto |> GenericChart.toChartHTML
-(***include-it-raw***)
\ No newline at end of file
+(***include-it-raw***)
+
+
+(**
+###SAM
+SAM (Significance Analysis of Microarrays) is a method developed to overcome multiple testing problems, for example in microarray experiments.
+SAM, short for Significance Analysis of Microarrays, is a statistical analysis created, but not restricted for, microarray analysis. 
+It serves as a blue print for any permutation test.
+Therefore, high throughput experiments can be analysed using a combined permutation-bootstrap-method. For more in depth information see the [SAM BlogPost](https://csbiology.github.io/CSBlog/posts/7_SAM.html).
+*)
+
+(** 
+Data: 
+To use SAM, data need to be in the format (string*float[])[]), with string being the name and float array being the repeated measurements. One way of achieving this is the following data preparation:
+<center><img style="max-width:40%" src="../img/DataStructureSAM.jpeg"></img></center>
+*)
+(**
+Rows are indicated by the sample name, here 1 and 2, representing control and factor. Columns are indicated with the rowkey(A1), here the gene id. This can be saved as .txt. 
+The next step is to read in the data, e.g. via deedle, and to create a dataframe. The rows are indexed by the sample name and the rowkeys are extracted.
+*)
+
+let df:Frame<string,string> = 
+    Frame.ReadCsv(@"data/TestDataSAM.txt",hasHeaders=true,separators = "\t")
+    // here, the name of A1 is needed. 
+    |> Frame.indexRows "gene"
+
+// get Rowkeys as Array
+let rowheader :string[] = df.RowKeys |> Seq.toArray
+
+// to separate control and factor sets (sample1 and sample2, respectively), 
+// the datasets are chunked by the number of repeated measurements (here: triplicates -> chunkBySize 3).
+let (preData1,preData2) :float[][] * float [][]=  
+    df
+    |> Frame.getRows
+    |> Series.values
+    |> Seq.map (Series.values >> Seq.toArray >> Array.chunkBySize 3 >> fun x -> x.[0],x.[1])
+    |> Array.ofSeq
+    |> Array.unzip
+
+// After chunking, the datasets are separated by factor and the rowkey is added to the data for later identification.
+let tupel a b = (a,b)
+let data1 = Array.map2 tupel rowheader preData1 
+let data2 = Array.map2 tupel rowheader preData2 
+(**
+Optional: Data can be normalised by median centering using the following function:
+*)
+open SAM
+let corrected1 = 
+    let medCorrect = 
+        medianCentering data1 
+    Array.map2 tupel rowheader medCorrect
+let corrected2 = 
+    let medCorrect = 
+        medianCentering data2
+    Array.map2 tupel rowheader medCorrect
+ 
+(**
+This calculates the median of a column and subtracts it from each value in this column. 
+With data1 and data2 , or corrected1 and corrected2, SAM can be performed.
+*)
+let res = FSharp.Stats.Testing.SAM.twoClassUnpaired 100 0.05 data1 data2 (System.Random(1234))
```

Comment:
The calculation fails if there is no significant down- or upregulated element within the data! In your test-dataset copy the 2 of the first class columns to also be the second class columns and set the pvalue threshold to 0.0001. It seems somewhere an array is called that is empty.